main.go

```diff
@@ -351,3 +319,97 @@ func uploadFile(uploadURL string, archiveFilePath string) error {
 
 	return nil
 }
+
+func getNewSuccessValue(currentOverallSuccess bool, stepWasSuccessful bool, wasLastStep bool, includedFlakyRetries bool) bool {
+	// Being overly cautious, by only setting success to true if it's the last try and flaky tests were enabled
+	// Doing this simply because there could be unaccounted for scenarios where it's not desirable to set successful to true
+	if !stepWasSuccessful {
+		return false
+	}
+	if stepWasSuccessful && wasLastStep && includedFlakyRetries {
+		return true
+	}
+	return currentOverallSuccess
+}
```

Comment:
I think we can get rid of this logic if we do the step preprocessing a bit different.

```
steps := responseModel.Steps
if configs.FlakyTestAttempts > 0 {
	steps, err = filteredSteps(steps)
	if err != nil {
		failf("Failed to filter latest steps, error: %s", err)
	}
}

for _, step := range steps {
...
}

func filteredSteps(steps []*toolresults.Step) ([]*toolresults.Step, error) {
	latestSteps := make(map[string]*toolresults.Step)

	for _, step := range steps {
		bytes, err := json.Marshal(step.DimensionValue)
		if err != nil {
			return nil, err
		}

		key := string(bytes)

		if previousStep, ok := latestSteps[key]; ok {
			if previousStep.CompletionTime != nil && step.CompletionTime != nil {
				if previousStep.CompletionTime.Seconds < step.CompletionTime.Seconds {
					latestSteps[key] = step
				}
			} else {
				// This should not really happen, but there is always a possibility that the API forgets to send a time.
				// I can see two actions here.
				//
				// One is to throw an error and treat it as an invalid case. Logically finished steps should always have
				// a finished at time.
				//
				// The second one is to keep the one without a time as a separate, individual entry.
			}
		} else {
			latestSteps[key] = step
		}
	}

	values := make([]*toolresults.Step, 0, len(latestSteps))
	for _, step := range latestSteps {
		values = append(values, step)
	}

	return values, nil
}
```
If the flaky retries are turned on then we can first go over the steps and keep only the last run of each step. That way we do not need to keep track of overall success or if this is the last step or if flaky retries were turned on. We can execute the same logic as before just on a different step array if the flaky retry was turned on.